### DISCUSSION TITLE
Retrieve User's Chosen Wallet (metamask, coinbase, argent, etc.)

### DISCUSSION THREAD

**Discussion by remibeaufils** [WalletConnect/walletconnect-monorepo#4027] [Developer Support]
**Issue:**
I'm facing a challenge in identifying which wallet the user has chosen to connect. I've explored the documentation and codebase but couldn't find a solution.

**Description:**
In the context of our application, we use web3modal, and it would be immensely beneficial to retrieve information about the user's chosen wallet during the connection process. This information is crucial for improving user experience and our analytics.

**Request:**
I'm reaching out to request assistance or guidance on how to implement this functionality or if there's a way to access the selected wallet information through the existing API.

Thank you for your assistance!

**Comment by glitch-txs** [comment]
Hi! You can find the wallet metadata information in the WalletConnect provider once the user is connected.

```ts
provider.session.peer.metadata
```

You can get the wallet provider from Web3Modal SDK in different ways depending on the peer ethereum library you're using (ethers or wagmi)

**Wagmi**

```ts
import { useAccount } from 'wagmi'

const { connector } = useAccount()

const walletInfo = (await connector?.getProvider()).session?.peer
```

**Ethers**

```ts
import { useWeb3ModalProvider } from '@web3modal/ethers/react'

const { walletProvider } = useWeb3ModalProvider()
walletProvider?.session?.peer
```

**Comment by remibeaufils** [comment]
Thank you for your answer.
However I just tested with Wagmi and it returns undefined :(

```ts
const walletInfo = (await connector?.getProvider()).session?.peer

// --> session is undefined
```

Do you have any idea why?

**Context:**
I tried using Chrome on macOS and connected with Metamask.
I'm using these package versions:
```json
"@web3modal/wagmi": "^3.4.0",
"viem": "^1.19.13",
"wagmi": "^1.4.10",
```

**Reply by glitch-txs** [reply]
the `session` object will only exists when you're are connected with WalletConnect

**Reply by glitch-txs** [reply]
for other wallets you can check the id/name

**Reply by remibeaufils** [reply]
Thank you for your answer.
However I just tested and `id` and `name` are undefined also

```ts
import { useAccount } from 'wagmi'

const { connector } = useAccount()
const provider = await connector?.getProvider()
provider.session // undefined
provider.name // undefined
provider.id // undefined
```

Would you have another suggestion?

**Reply by DanilKasianenko** [reply]
Yea I have the same all data `undefined`

```tsx
  const { address, connector } = useAccount();

  const getProvider = async () => {
    const walletInfo = await connector?.getWalletClient();
    const account = await connector?.getAccount();
    console.log("walletInfo", walletInfo); // undefined
    console.log("account", account); // undefined
  };
```

**Reply by glitch-txs** [reply]
it works for me

added this snippet in Next
```ts
  useEffect(() => {
    async function getWalletData() {
      if (!connector) return;
      const provider = await connector.getProvider();
      console.log(provider.session);
    }
    getWalletData();
  }, [connector]);
```
![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/66949816/c4d52f3a-a028-496a-991b-a576b2a03a20)

**Reply by remibeaufils** [reply]
Thank you for your answer @glitch-txs.
I just tested, and indeed the session object exists when I'm connected with WalletConnect 🙏

To come back to my initial need, I use web3modal and I need to retrieve information about the user's chosen wallet during the connection process. This information is crucial for improving user experience and our analytics.

Your solution works in case the user connects through WalletConnect.
But not when he connects directly with a wallet, eg MetaMask.

In this case, I tested again and `session` and `id` and `name` are always undefined.

Would you have another suggestion?

**Reply by glitch-txs** [reply]
I think we could build a hook for this, could you open an issue in Web3Modal repository for this? we could abstract WalletConnect & EIP-6963 connections into a hook that returns the wallet information or even in the events hook

**Reply by remibeaufils** [reply]
Absolutely! I've opened the requested issue in the Web3Modal repository. You can find it [here](https://github.com/WalletConnect/web3modal/issues/1665). Looking forward to your feedback!

**Reply by DanilKasianenko** [reply]
What packages versions do you have? 

Because of code sniped don't work on my project :(
